notebooks/ir_ima_visualization/ima_visualization_and_differencing.py

```diff
@@ -0,0 +1,327 @@
+import numpy as np
+from astropy.io import fits
+import matplotlib.pyplot as plt
+from ginga.util.zscale import zscale
+import matplotlib.patheffects as path_effects
+
+def read_wfc3(filename):
+    '''
+    Read a full-frame IR image and return the datacube plus integration times for each read.
+    Requires the path to a RAW full-frame IR image fits file (filename.)
+
+    Parameters
+    ----------
+    filename : str
+        Path to a RAW full-frame IR image fits file.
+
+    Returns
+    -------
+    cube : array-like
+        1024x1024xNSAMP datacube of the IR image in ascending time order, 
+        where NSAMP is the number of samples taken.
+  
+    integ_time : array-like
+        Integration times associated with the datacube in ascending order.
+        
+    '''
+    with fits.open(filename) as f:
+        hdr = f[0].header
+        NSAMP = hdr['NSAMP']
+        hdr1 = f[1].header
+        cube = np.zeros((hdr1['NAXIS1'],hdr1['NAXIS2'],NSAMP), dtype = float)
+        integ_time = np.zeros(shape = (NSAMP))
+
+        for i in range(1, NSAMP+1):
+            cube[:,:,i-1] = f[('SCI', i)].data
+            integ_time[i-1] = f[('TIME', i)].header['PIXVALUE']
+
+        cube = cube[:,:,::-1]
+        integ_time = integ_time[::-1]
+    return cube, integ_time
+
+
+
+def compute_diff_imas(cube, integ_time, diff_method):
+    
+    '''
+    Compute the difference in signal between reads of a WFC3 IR IMA file.
+    
+    Parameters
+    ----------
+    cube : array-like
+       1024x1024xNSAMP datacube of the IR image in ascending time order, 
+       where NSAMP is the number of samples taken.
+
+    integ_time : array-like
+       Integration times associated with the datacube in ascending order.
+
+    diff_method: str
+       The method of finding the difference between reads. 
+       Either "instantaneous" or "cumulative".
+           
+    Returns
+    --------
+    diff : array-like
+        1024x1024x(NSAMP-1) datacube of the differebce between IR IMA reads in ascending time order, 
+        where NSAMP is the number of samples taken.
+        
+    '''
+    
+    if diff_method == 'instantaneous':
+        ima_j = cube[:, :, 1:]
+        ima_j_1 = cube[:,:,0:-1]
+        diff = ((ima_j*(integ_time[1:]-integ_time[0]))-\
+        (ima_j_1*(integ_time[0:-1]-integ_time[0])))/(integ_time[1:]-integ_time[0:-1])
+        
+    elif diff_method == 'cumulative':
+        diff = cube[:,:,0:-1] - cube[:,:,1:]
```

Comment:
Just to be crystal clear, are the ima reads supposed to be subtracted in reverse order? I'm sorry if that was already mentioned in the text but I'm curious as to why for example you would perform SCI[15] - SCI[14] if SCI[14] came _after_ SCI[15] (hence reversed time order) and if for instantaneous it's subtracted in time order (pseudo SCI[14] - SCI[15]).

Reply:
@annierose3 Can you address this real quick?

Reply:
also, the instantaneous method is psuedo 14-15, but since it is ~derivative-like~ more than just subtraction, it isn't a negative difference in rate and works fine.